File: services/context.py
"""
Context 工程：
- analyze：发送前看 messages 里有什么（按 role 统计）
- trim：发送前裁剪，防止 history / tool 结果过长
- 发送后用 resp.usage 看真实 token（在 agent_loop 里打）
"""

MAX_TOOL_CONTENT_CHARS = 2000
KEEP_LAST_MESSAGES = 12


def message_chars(msg: dict) -> int:
    n = len(msg.get("content") or "")
    for tc in msg.get("tool_calls") or []:
        fn = tc.get("function") or {}
        n += len(fn.get("name") or "")
        n += len(fn.get("arguments") or "")
    return n


def analyze_messages(messages: list[dict]) -> dict:
    by_role: dict[str, dict] = {}
    for msg in messages:
        role = msg.get("role", "unknown")
        if role not in by_role:
            by_role[role] = {"count": 0, "chars": 0}
        by_role[role]["count"] += 1
        by_role[role]["chars"] += message_chars(msg)
    return {
        "total_messages": len(messages),
        "total_chars": sum(message_chars(m) for m in messages),
        "by_role": by_role,
    }


def format_analysis(analysis: dict) -> str:
    lines = [
        f"messages={analysis['total_messages']}, chars≈{analysis['total_chars']}",
    ]
    for role, data in analysis["by_role"].items():
        lines.append(f"  {role}: count={data['count']}, chars={data['chars']}")
    return "\n".join(lines)


def truncate_text(text: str, max_chars: int) -> str:
    if len(text) <= max_chars:
        return text
    half = max_chars // 2 - 24
    left = text[:half]
    right = text[-half:]
    return f"{left}\n... [truncated, total {len(text)} chars] ...\n{right}"


def shrink_tool_content(content: str) -> str:
    if not content:
        return content
    lower = content.lower()
    if any(k in lower for k in ("error", "fail", "✗")):
        lines = content.splitlines()
        picked: list[str] = []
        picked.extend(lines[:5])
        for line in lines:
            if any(k in line.lower() for k in ("error", "fail", "✗")):
                picked.append(line)
        picked.extend(lines[-5:])
        seen: set[str] = set()
        unique: list[str] = []
        for line in picked:
            if line not in seen:
                seen.add(line)
                unique.append(line)
        return truncate_text("\n".join(unique), MAX_TOOL_CONTENT_CHARS)
    return truncate_text(content, MAX_TOOL_CONTENT_CHARS)


def _copy_and_shrink_tools(messages: list[dict]) -> list[dict]:
    out: list[dict] = []
    for msg in messages:
        m = msg.copy()
        if m.get("role") == "tool" and isinstance(m.get("content"), str):
            m["content"] = shrink_tool_content(m["content"])
        out.append(m)
    return out


def trim_messages(messages: list[dict]) -> list[dict]:
    if not messages:
        return []
    system_msgs = [m for m in messages if m.get("role") == "system"]
    rest_msgs = [m for m in messages if m.get("role") != "system"]
    if len(rest_msgs) > KEEP_LAST_MESSAGES:
        rest_msgs = rest_msgs[-KEEP_LAST_MESSAGES:]
    return system_msgs + _copy_and_shrink_tools(rest_msgs)

File: apps/server/src/context.ts
const MAX_TOOL_CONTENT_CHARS = 2000;
const KEEP_LAST_MESSAGES = 12;

export type ChatMessage = Record<string, unknown>;

function messageChars(msg: ChatMessage): number {
  let n = String(msg.content ?? "").length;
  const toolCalls = (msg.tool_calls as unknown[]) ?? [];
  for (const tc of toolCalls) {
    const row = tc as { function?: { name?: string; arguments?: string } };
    n += (row.function?.name ?? "").length;
    n += (row.function?.arguments ?? "").length;
  }
  return n;
}

function truncateText(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;
  const half = Math.floor(maxChars / 2) - 24;
  const left = text.slice(0, half);
  const right = text.slice(-half);
  return `${left}\n... [truncated, total ${text.length} chars] ...\n${right}`;
}

function shrinkToolContent(content: string): string {
  if (!content) return content;
  const lower = content.toLowerCase();
  if (["error", "fail", "✗"].some((k) => lower.includes(k))) {
    const lines = content.split("\n");
    const picked: string[] = [];
    picked.push(...lines.slice(0, 5));
    for (const line of lines) {
      if (["error", "fail", "✗"].some((k) => line.toLowerCase().includes(k))) {
        picked.push(line);
      }
    }
    picked.push(...lines.slice(-5));
    const seen = new Set<string>();
    const unique: string[] = [];
    for (const line of picked) {
      if (!seen.has(line)) {
        seen.add(line);
        unique.push(line);
      }
    }
    return truncateText(unique.join("\n"), MAX_TOOL_CONTENT_CHARS);
  }
  return truncateText(content, MAX_TOOL_CONTENT_CHARS);
}

function copyAndShrinkTools(messages: ChatMessage[]): ChatMessage[] {
  return messages.map((msg) => {
    if (msg.role === "tool" && typeof msg.content === "string") {
      return { ...msg, content: shrinkToolContent(msg.content) };
    }
    return { ...msg };
  });
}

export function trimMessages(messages: ChatMessage[]): ChatMessage[] {
  if (!messages.length) return [];
  const systemMsgs = messages.filter((m) => m.role === "system");
  let restMsgs = messages.filter((m) => m.role !== "system");
  if (restMsgs.length > KEEP_LAST_MESSAGES) {
    restMsgs = restMsgs.slice(-KEEP_LAST_MESSAGES);
  }
  return [...systemMsgs, ...copyAndShrinkTools(restMsgs)];
}

export function analyzeMessages(messages: ChatMessage[]) {
  const byRole: Record<string, { count: number; chars: number }> = {};
  for (const msg of messages) {
    const role = String(msg.role ?? "unknown");
    if (!byRole[role]) byRole[role] = { count: 0, chars: 0 };
    byRole[role].count += 1;
    byRole[role].chars += messageChars(msg);
  }
  return {
    total_messages: messages.length,
    total_chars: messages.reduce((sum, m) => sum + messageChars(m), 0),
    by_role: byRole,
  };
}

export function formatAnalysis(analysis: ReturnType<typeof analyzeMessages>): string {
  const lines = [
    `messages=${analysis.total_messages}, chars≈${analysis.total_chars}`,
  ];
  for (const [role, data] of Object.entries(analysis.by_role)) {
    lines.push(`  ${role}: count=${data.count}, chars=${data.chars}`);
  }
  return lines.join("\n");
}
